core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala

```diff
@@ -315,130 +316,137 @@ trait SubmissionMonitor extends FutureSupport with LazyLogging with RawlsInstrum
-    * the WorkflowRecords in ExecutionServiceStatus response have not been saved to the database but have been updated with their status from Cromwell.
+    * once all the execution service queries have completed this function is called to handle the responses
+    * the WorkflowRecords in ExecutionServiceStatus response have not been saved to the database but
+    * have been updated with their status from Cromwell.
     *
-    * @param response
-   * @param executionContext
-   * @return
-   */
+    * @param response all workflows in this submission whose statuses are different in Cromwell than in Rawls. If a
+    *                 workflow has outputs in Cromwell, those outputs are included here.
+    * @param executionContext
+    * @return
+    */
   def handleStatusResponses(
     response: ExecutionServiceStatusResponse
   )(implicit executionContext: ExecutionContext): Future[StatusCheckComplete] =
+    // start a trace
     trace("SubmissionMonitorActor.handleStatusResponses") { rootSpan =>
       rootSpan.tracingSpan.foreach { s =>
         s.putAttribute("submissionId", OpenCensusAttributeValue.stringAttributeValue(submissionId.toString))
         s.putAttribute("workspaceNamespace", OpenCensusAttributeValue.stringAttributeValue(workspaceName.namespace))
         s.putAttribute("workspaceName", OpenCensusAttributeValue.stringAttributeValue(workspaceName.name))
       }
 
+      // log any failures.
       response.statusResponse.collect { case Failure(t) => t }.foreach { t =>
         logger.error(s"Failure monitoring workflow in submission $submissionId", t)
       }
 
-      // all workflow records in this status response list
-      val allWorkflows = response.statusResponse.collect { case Success(Some((aWorkflow, _))) =>
-        aWorkflow
-      }
-
-      // just the workflow records in this response list which have outputs
-      val workflowsWithOutputs = response.statusResponse.collect { case Success(Some((workflowRec, Some(outputs)))) =>
-        (workflowRec, outputs)
-      }
-
-      def markWorkflowsFailed(workflows: Seq[WorkflowRecord], fatal: RawlsFatalExceptionWithErrorReport) = {
-        logger.error(
-          s"Marking ${workflows.length} workflows as failed handling outputs in $submissionId with user-visible reason ${fatal.toString})"
-        )
-        datasource.inTransaction { dataAccess =>
-          DBIO.sequence(workflows map { workflowRecord =>
-            dataAccess.workflowQuery.updateStatus(workflowRecord, WorkflowStatuses.Failed) andThen
-              dataAccess.workflowQuery.saveMessages(Seq(AttributeString(fatal.toString)), workflowRecord.id)
-          })
+      // partition workflows into two variables: those workflows that do have
+      // outputs vs. those workflows that do not have outputs. This allows us to process
+      // the easier workflows first (those without outputs)
+      val (yesOutputs, noOutputs) = response.statusResponse
```

Comment:
I'm not sure we have more than a trivial number of no-output workflows, I've thought about eliminating this partition in the past. Seems like the tracing would help answer that question.

Reply:
I ran a trivial calculate-an-md5 workflow in my bee, and the logs showed:
```
[INFO] [18:28:23.917] [rawls-submission-monitor-dispatcher-54] o.b.d.r.j.SubmissionMonitorActor - will process 18 workflow(s) without outputs and 34 workflow(s) with outputs for submission 754c7205-2808-40f1-b3e5-30e504704072
[INFO] [18:29:30.214] [rawls-submission-monitor-dispatcher-49] o.b.d.r.j.SubmissionMonitorActor - will process 10 workflow(s) without outputs and 56 workflow(s) with outputs for submission 754c7205-2808-40f1-b3e5-30e504704072
[INFO] [18:30:34.914] [rawls-submission-monitor-dispatcher-54] o.b.d.r.j.SubmissionMonitorActor - will process 0 workflow(s) without outputs and 10 workflow(s) with outputs for submission 754c7205-2808-40f1-b3e5-30e504704072
```

I wonder if - outside of this PR - there's an enhancement where `queryExecutionServiceForStatus` doesn't return any workflows which are not in a terminal status. It may be that `handleStatusResponses` is processing workflows it doesn't need to.

Reply:
See also the output captured in this comment: https://github.com/broadinstitute/rawls/pull/2680#issuecomment-1879303953. The most likely case for workflows without outputs are those that have just transitioned from Submitted -> Running … by using the `.partition` here, we prioritize marking those workflows as Running in rawls as soon as possible, before writing outputs for any workflows (in this batch) that have completed.